crates/compiler/mono/src/ir.rs

```diff
@@ -7411,6 +7411,28 @@ fn store_pattern_help<'a>(
             // do nothing
             return StorePattern::NotProductive(stmt);
         }
+        As(subpattern, symbol) => {
+            let stored_subpattern =
+                store_pattern_help(env, procs, layout_cache, subpattern, outer_symbol, stmt);
+
+            let mut stmt = match stored_subpattern {
+                StorePattern::Productive(stmt) => stmt,
+                StorePattern::NotProductive(stmt) => stmt,
+            };
+
+            // An identifier in a pattern can define at most one specialization!
+            // Remove any requested specializations for this name now, since this is the definition site.
+            let specialization_symbol = procs
+                .symbol_specializations
+                .remove_single(*symbol)
+                // Can happen when the symbol was never used under this body, and hence has no
+                // requested specialization.
+                .unwrap_or(*symbol);
+
+            substitute_in_exprs(env.arena, &mut stmt, specialization_symbol, outer_symbol);
```

Comment:
@ayazhafiz does this reasoning make sense for the identifier introduced by `as`? it's the same as for normal `Identifier` patterns, which I think is correct.

Reply:
Yeah that seems right.
I'm working on weakening let-generalization so when that pans out, we'll be able to rip out all of this (finally)